### DISCUSSION TITLE
prometheus: fatal error: runtime: out of memory

### DISCUSSION THREAD

**Discussion by asaf1602** [prometheus/prometheus#10178] [Q&A] [answered]
**What did you do?**
We installed prometheus v2.25.0 on VM with 4 cores/16GB memory
The process failed with OOM few time. For restart it we must delete all data under the data dir (--storage.tsdb.path) 
The total size in data dir is ~4GB 

**What did you expect to see?**
Prometheus will run
**What did you see instead? Under which circumstances?**
The memory "Memory in Off-Heap" is grow all the time until process is failed with OOM
**Environment**

* System information:
        infadm@monitor01.ocs.pet:infadm->uname -srm
        Linux 3.10.0-1127.el7.x86_64 x86_64
        infadm@monitor01.ocs.pet:infadm->cat /proc/cpuinfo | grep processor | wc -l
        4
        infadm@monitor01.ocs.pet:infadm->free -h
                      total        used        free      shared  buff/cache   available
        Mem:            15G        1.5G         12G        898M        1.1G         12G
        Swap:            0B          0B          0B

* Prometheus version:
        infadm@monitor01.ocs.pet:infadm->/usr/local/bin/prometheus --version
        prometheus, version 2.25.0 (branch: HEAD, revision: a6be548dbc17780d562a39c0e4bd0bd4c00ad6e2)
          build user:       root@615f028225c9
          build date:       20210217-14:17:24
          go version:       go1.15.8
          platform:         linux/amd64
        infadm@monitor01.ocs.pet:infadm->_
* Alertmanager version:
Not installed/used, yet, in this env.

* Prometheus configuration file:
global:
  scrape_interval: 30s
  scrape_timeout: 30s
  evaluation_interval: 30s
alerting:
  alertmanagers:
  - scheme: http
    timeout: 10s
    api_version: v1
    static_configs:
    - targets: []
rule_files:
- /amdocs/TCMonitoring/AlertRules/AlertRules*.yml
scrape_configs:
- job_name: prometheus
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets:
    - localhost:9090
- job_name: TCExporters_nodes
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  file_sd_configs:
  - files:
    - /amdocs/TCMonitoring/scrapeFiles/tcmon_node*.json
    refresh_interval: 5m
- job_name: TCExporters_jmx
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  file_sd_configs:
  - files:
    - /amdocs/TCMonitoring/scrapeFiles/tcmon_jmx*.json
    refresh_interval: 5m
- job_name: TCExporters_PG
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  file_sd_configs:
  - files:
    - /amdocs/TCMonitoring/scrapeFiles/tcmon_pg*.json
    - /amdocs/TCMonitoring/scrapeFiles/tcmon_patroni*.json
    refresh_interval: 5m
- job_name: TCExporters_airflow
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /admin/metrics
  scheme: http
  file_sd_configs:
  - files:
    - /amdocs/TCMonitoring/scrapeFiles/*airflow*.json
    refresh_interval: 5m
- job_name: TCFederate_MS360
  honor_labels: true
  honor_timestamps: true
  params:
    match[]:
    - '{job="asm-services-cloud-service", namespace="pet-ms360-ocs"}'
    - '{job="asm-services-cloud-service-custom", namespace="pet-ms360-ocs"}'
    - '{job="fndsec-federation", namespace="pet-ms360-ocs"}'
    - '{job="fndsec-pki-operator", namespace="pet-ms360-ocs"}'
    - '{job="fndsec-sso-provider", namespace="pet-ms360-ocs"}'
    - '{job="fndsec-sts", namespace="pet-ms360-ocs"}'
    - '{job="fndsec-sts-trust-reg", namespace="pet-ms360-ocs"}'
    - '{job=~"charging-provisioning-gateway.*", namespace="pet-ms360-ocs"}'
    - '{job="prepaybalance-management-service", namespace="pet-ms360-ocs"}'
    - '{job=~"usage.*management", namespace="pet-ms360-ocs"}'
    - '{job=~"reratetask.*service", namespace="pet-ms360-ocs"}'
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /federate
  scheme: http
  static_configs:
  - targets:
    - 10.172.142.21:9090
    labels:
      siteID: "1"
- job_name: TCFederate_TCK8S
  honor_labels: true
  honor_timestamps: true
  params:
    match[]:
    - '{job="kube-state-metrics"}'
    - '{job="kubelet", namespace="pet-k8s-ocs"}'
    - '{job=~".*-srv.*", namespace="pet-k8s-ocs"}'
    - '{job="node-exporter",prometheus_replica="prometheus-k8s-1"}'
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /federate
  scheme: http
  static_configs:
  - targets:
    - 10.172.142.21:9090
    labels:
      siteID: "1"
* Alertmanager configuration file:
N/A
insert configuration here (if relevant to the issue)
N/A

* Logs:
[prometheus-Dec-29.log.gz](https://github.com/prometheus/prometheus/files/7792062/prometheus-Dec-29.log.gz) - All prometheus messages from /var/log/messages
![Promtheus-Memory_2](https://user-images.githubusercontent.com/47528794/147734231-5dfac324-9b62-4793-b7ca-ba4768c246bc.gif) Metric go_memstats_heap_alloc_bytes from prometheus GUI
![Promtheus-Memory](https://user-images.githubusercontent.com/47528794/147734234-4fa8f9fd-14f3-4348-a478-2680be64a911.png) - Go Metrics Dashboard
![Promtheus-Memory_1](https://user-images.githubusercontent.com/47528794/147734240-4a5c612a-2a19-4b49-b4b1-9e90e7adb539.png) - Prometheus Benchmark - 2.17.x Dashboard
[Promtheus_CommandLineFlags_and_RuntimeFlags.txt](https://github.com/prometheus/prometheus/files/7792064/Promtheus_CommandLineFlags_and_RuntimeFlags.txt)

**Comment by roidelapluie** [comment]
From the logs it looks like something is happening during compation. Could you try to upgrade to the latest release and see if the problem still occurs?

**Comment by asaf1602** [comment]
Hi,

I cant upgrade the version (production ...)
But, at 123-Jan we disable the THP (Transparent Huge Pages) and from this time we didnt see any out of memory issue. As u can see in the Graph, memory used behaved as expected.
Do u know about any limitation related to THC? 

Asaf

[Promtheus-Memory_Metric-go_memstats_heap_alloc_bytes_Jan12ToNow](https://user-images.githubusercontent.com/47528794/149649222-172fa68c-98d3-48c7-9d9b-9dfb7e000eda.gif)

**Comment by asaf1602** [comment]
Hi,

Prometheus failed again :( 
Attached the prometheus messages from /var/log/messages 

Any suggestion will be helpful 
10x,
Asaf
[Prometues.log](https://github.com/prometheus/prometheus/files/7880040/Prometues.log)

**Comment by asaf1602** [comment]
GM,
Someone?

**Comment by roidelapluie** [comment]
it would be helpful if we could get a screenshot of your TSDB stats page when Prometheus is up. Again, I also recommend you to upgrade to the latest stable release.

**Comment by asaf1602** [comment]
Hi
10x for your response
I didn't find the tsdb in my installation.

I found it in https://github.com/prometheus-junkyard/tsdb
But when I download the zip file and open it, I didn't found 'tsdb'.
Can you help to find it?

Asaf

**Comment by roidelapluie** [comment]
I mean, the "tsdb status" page in the Prometheus UI.

**Comment by asaf1602** [comment]
Hi,

See attached.

![P8S_TSDB_1-Screenshot-2022-01-18](https://user-images.githubusercontent.com/47528794/149940576-a2176e48-975f-407b-8630-69d7d8e3abec.gif)
![P8S_TSDB_2-Screenshot-2022-01-18](https://user-images.githubusercontent.com/47528794/149940578-3e7d7627-51ee-4072-aed6-2c212c6dcb88.gif)

**Comment by asaf1602** [comment]
![Go-Dashboad-now-210m_now](https://user-images.githubusercontent.com/47528794/149942761-93937604-6431-4b42-9403-0a682440bd98.gif)

**Comment by asaf1602** [comment]
![Go-Dashboad-Last1H](https://user-images.githubusercontent.com/47528794/149942877-1127ff1d-88c7-458b-9ced-03411789ae6b.gif)

**Comment by roidelapluie** [comment]
Thank you. It seems that your memory issue is due to your usage of Prometheus. From what I see, you have unbound cardinality in your queryid and short_query labels, therefore you seem to be tracking events rather than metrics.

I don't think there is more we can do in this issue, but here is a presentation about cardinality and Prometheus that could help you: https://promcon.io/2019-munich/talks/containing-your-cardinality/

**Comment by SuperQ** [comment, marked as answer]
I'd recommend looking at [this queries.yaml](https://gitlab.com/gitlab-cookbooks/gitlab-exporters/-/blob/b33a2d7090b668205e94616df7c793a472d0d88e/templates/postgres_exporter/queries.yaml.erb#L282-328) to reduce your `pg_stat_statements` cardinality.

**Comment by asaf1602** [comment]
Hi
10x for your comments 
I will check the DBA metrics

**Comment by asaf1602** [comment]
GM,
10x for your help. temporary I removed the scraping from PG. Prometheus is much stable.
